Safe processing of darkroom thumb

jenshannoschwalm opened the pull request:
As the preview pipe can possibly change while displaying the thumb-in-development there could be wrong dimensions possibly resulting in out-of-buffer memcpy.

Instead of doing "complicated things" we just do the whole processing while the preview pipe has been locked. No responsiveness- or performance penalty as this is done asynchronously anyway.

Likely Fixes #16785

ralfbrown commented on src/dtgtk/thumbnail.c:
```diff
@@ -646,16 +646,20 @@ static gboolean _event_image_draw(GtkWidget *widget,
+      const size_t bbufsize = sizeof(uint8_t) * 4 * buf_width * buf_height;
+      uint8_t *rgbbuf = dt_alloc_align_uint8(bbufsize);
+
+      memcpy(rgbbuf, dev->preview_pipe->backbuf, bbufsize);
```
It's safe to unlock the mutex immediately after the memcpy, since no more references are made to the pipe's data, only local copies.  I've just checked, and the only place that points `backbuf` at a new buffer also sets `backbuf_width` and `backbuf_height` in the same critical section, so there's no possibility of seeing inconsistent data here.

jenshannoschwalm replied:
Right, cross checked too.

jenshannoschwalm commented:
fixed as suggested, corrected comments in code & commit message and force pushed.

Release note could be: Fixed thumbnail updating while changing image while in darkroom

ralfbrown commented:
I recompiled after manually moving the mutex lock call before the lines which read the size (I didn't want to spend the time figuring out how to merge a PR into my local repo) and just cranked through editing 190 images in an hour with the filmstrip enabled without a crash.  Looks like this PR does fix the crashing.

wpferguson commented:
> I didn't want to spend the time figuring out how to merge a PR into my local repo

@ralfbrown Here's a script to do it.  Give it a PR number and a branch to put it on and it does the work.  

[test_dt_pr.sh.txt](https://github.com/darktable-org/darktable/files/15322611/test_dt_pr.sh.txt)

jenshannoschwalm commented on src/dtgtk/thumbnail.c:
```diff
@@ -692,16 +696,16 @@ static gboolean _event_image_draw(GtkWidget *widget,
         {
           cairo_save(cr2);
           cairo_scale(cr2, 1.0f/scale, 1.0f/scale);
+          cairo_surface_flush(thumb->img_surf);
```
Nope, not directly related. Stumbled across this (plus the other pr making sure we don't use data outside of safe range) while reading this part of code ...

About the flush
***
Do any pending drawing for the surface and also restore any temporary modifications cairo has made to the surface's state. This function must be called before switching from drawing on the surface with cairo to drawing on it directly with native APIs, or accessing its memory outside of Cairo. If the surface doesn't support direct access, then this function does nothing.
***
from the cairo docs. Not absolutely sure ...

jenshannoschwalm replied:
Sorry for the noise, i checked the focus peaking code agin and there is nothing done that would require such protections. (If there were such ops, we would have to do cairo_surface_mark_dirty() as the second op but here we need nothing like that.

ralfbrown commented:
160 images edited with the PR applied with no crashes from the backbuf changing.  Unfortunately, two crashes from other causes (once merging histories while pasting and once in `dt_mipmap_cache_get_with_caller` at
```
        dt_image_t DT_ALIGNED_ARRAY buffered_image;
        const dt_image_t *cimg = dt_image_cache_get(darktable.image_cache, imgid, 'r');
        buffered_image = *cimg;              <-- crash here
 ```

jenshannoschwalm commented:
Force pushed after removing unnecessary stuff, so ready to merge from my side.

No idea yet about the issue mentioned by @ralfbrown yet ...
